h/admin/views/users.py

```diff
@@ -69,23 +65,59 @@ def users_activate(request):
                                     _query=(('username', user.username),)))
 
 
+@view_config(route_name='admin_users_rename',
+             request_method='POST',
+             permission='admin_users')
+def users_rename(request):
+    user = _form_request_user(request)
+
+    if user is None:
+        return httpexceptions.HTTPFound(
+            location=request.route_path('admin_users'))
+
+    old_username = user.username
+    new_username = request.params.get('new_username')
+
+    try:
+        # FIXME (1) - This assignment validates the username length but not
+        # the character set or any other restrictions.
+        user.username = new_username
+
+        # FIXME (2) - We need to trigger re-indexing of any of the user's
+        # annotations here and ensure that annotation permissions are updated
+        # as a result
+
+        # FIXME (3) - Is there already a service layer that this logic can go
+        # into? If not, should we create one and move the user
+        # rename/activate/delete logic there now?
```

Comment:
All of the business logic should probably run in a background worker. We can ensure that the user's username change and the annotations userid change happens at the same time (yay for transactional databases!), and then reindex their annotations (`h.api.search.index.BatchIndexer` can take a list of ids).
As with our current manual solution, there will be a certain time period when searching for the new username will not result in any, or not all annotations of the users, until the search index has been updated.

The problem with putting everything in a background worker is that we can't easily communicate when it is done since it depends on how much is already in the celery queue.

The code itself can still live in a service, but should be called in a celery worker in my opinion, to ensure that we don't run into any timeouts that could leave the index in a bad state (the database should be fine, again, yay for transactions!).